rust/src/sans_io/read.rs

```diff
@@ -0,0 +1,1014 @@
+            compressed: RWBuf::default(),
+            data_section_hasher: if options.validate_data_section_crc {
+                None
+            } else {
+                Some(crc32fast::Hasher::new())
+            },
+            calculated_data_section_crc: None,
+            decompressors: HashMap::new(),
+            last_write: None,
+            at_eof: false,
+            options,
+            failed: false,
+        }
+    }
+
+    /// Constructs a linear reader that will iterate through all records in a chunk.
+    pub(crate) fn for_chunk(header: ChunkHeader) -> McapResult<Self> {
+        let mut result = Self::new_with_options(
+            LinearReaderOptions::default()
+                .with_skip_end_magic(true)
+                .with_skip_start_magic(true)
+                .with_validate_chunk_crcs(true),
+        );
+        result.currently_reading = Record;
+        result.from = Chunk(ChunkState {
+            decompressor: result.get_decompressor(&header.compression)?,
+            hasher: Some(crc32fast::Hasher::new()),
+            crc: header.uncompressed_crc,
+            compressed_remaining: header.compressed_size,
+            padding_after_compressed_data: 0,
+        });
+        Ok(result)
+    }
+
+    fn get_decompressor(&mut self, name: &str) -> McapResult<Option<Box<dyn Decompressor>>> {
+        if let Some(decompressor) = self.decompressors.remove(name) {
+            return Ok(Some(decompressor));
+        }
+        match name {
+            #[cfg(feature = "zstd")]
+            "zstd" => Ok(Some(Box::new(zstd::ZstdDecoder::new()))),
+            #[cfg(feature = "lz4")]
+            "lz4" => Ok(Some(Box::new(lz4::Lz4Decoder::new()?))),
+            "" => Ok(None),
+            _ => Err(McapError::UnsupportedCompression(name.into())),
+        }
+    }
+    fn return_decompressor(&mut self, mut decompressor: Box<dyn Decompressor>) -> McapResult<()> {
+        decompressor.reset()?;
+        self.decompressors
+            .insert(decompressor.name().into(), decompressor);
+        Ok(())
+    }
+
+    /// Yields the next action the caller should take to progress through the file.
+    pub fn next_action(&mut self) -> Option<McapResult<ReadAction>> {
+        self.next_action_inner().transpose()
+    }
+
+    fn next_action_inner(&mut self) -> McapResult<Option<ReadAction>> {
+        if self.failed {
+            return Ok(None);
+        }
+
+        // process any newly-written data since the last next_action() call.
+        if let Some(written) = self.last_write.take() {
+            if written == 0 {
+                self.at_eof = true;
+            }
+
+            let tail = match &self.from {
+                File => self.uncompressed.tail(),
+                Chunk(state) => match &state.decompressor {
+                    Some(_) => self.compressed.tail(),
+                    None => self.uncompressed.tail(),
+                },
+            };
+            let written_region = &tail[..written];
```

Comment:
The comments I can find for tail() say "returns a mutable view of the un-written part of the buffer" – when did this become written? Does the comment need to be updated?

Reply:
This got written by the caller when they updated the InputBuf, but the caller did not update the start or end cursors then.